### PR TITLE
Add DispatchLoopBuilder and implement shutdown 

### PR DESCRIPTION
**Pull request by agunde406**
Before a DispatchLoop would need to be run in a separate thread
that splinterd would need to create. The loop would take an
Arc<AtomicBool> to signal shutdown and a recv to get messages from.

Now a DispatchLoopBuilder will take the dispatcher and start up its
own channel and thread on build. The DispatchLoop provides methods
to get a sender and a shutdown_signaler. This change also
required that the DispatchMessage got changed into a enum with
Message and Shutdown.

### CODE REVIEW

File: libsplinter/src/network/dispatch.rs [hunk 1]
```diff
@@ -301,112 +295,144 @@ impl<MT: Hash + Eq + Debug + Clone> HandlerWrapper<MT> {
...
-        &self.source_peer_id
+    pub fn build(mut self) -> Result<DispatchLoop<MT>, String> {
+        let (tx, rx) = channel();
+
+        let dispatcher = self
+            .dispatcher
+            .take()
+            .ok_or_else(|| "No dispatch provided".to_string())?;
+
+        let join_handle = std::thread::spawn(move || loop {
```
Comment (jsmitchell) [inline libsplinter/src/network/dispatch.rs:349]: use thread::Builder to give this thread a good name - https://doc.rust-lang.org/std/thread/#configuring-threads

File: splinterd/src/daemon.rs [hunk 2]
```diff
@@ -298,15 +307,16 @@ impl SplinterDaemon {
                                     }
                                 };
 
-                            let dispatch_msg = DispatchMessage::new(
+                            trace!("Received Message from {}: {:?}", message.peer_id(), msg);
```
Comment (jsmitchell) [inline splinterd/src/daemon.rs:310]: s/Message/message/

File: libsplinter/src/network/dispatch.rs [hunk 3]
```diff
@@ -301,112 +295,144 @@ impl<MT: Hash + Eq + Debug + Clone> HandlerWrapper<MT> {
...
+                        message_type,
+                        message_bytes,
+                        source_peer_id,
+                    }) => (message_type, message_bytes, source_peer_id),
+                    Ok(DispatchMessage::Shutdown) => {
+                        debug!("Received shutdown signal");
+                        break;
+                    }
+                    Err(RecvError) => {
+                        error!("Received Error from receiver");
```
Comment (jsmitchell) [inline libsplinter/src/network/dispatch.rs:362]: s/Error/error/

File: libsplinter/src/network/dispatch.rs [hunk 4]
```diff
@@ -301,112 +295,149 @@ impl<MT: Hash + Eq + Debug + Clone> HandlerWrapper<MT> {
...
+            error!("Unable to send shutdown signal to already shutdown dispatch loop");
         }
     }
+}
 
-    pub fn message_type(&self) -> &MT {
-        &self.message_type
+/// Errors that may occur during the operation of the Dispatch Loop.
+#[derive(Debug)]
+pub struct DispatchLoopError(String);
```
Comment (peterschwarz) [inline libsplinter/src/network/dispatch.rs:324]: Probably could implement `Error`

File: libsplinter/src/network/dispatch.rs [hunk 5]
```diff
@@ -301,112 +295,149 @@ impl<MT: Hash + Eq + Debug + Clone> HandlerWrapper<MT> {
...
+    pub fn build(mut self) -> Result<DispatchLoop<MT>, String> {
+        let (tx, rx) = channel();
+
+        let dispatcher = self
+            .dispatcher
+            .take()
+            .ok_or_else(|| "No dispatch provided".to_string())?;
+
+        let join_handle = std::thread::Builder::new()
+            .name("DispatchLoop".to_string())
```
Comment (peterschwarz) [inline libsplinter/src/network/dispatch.rs:350]: Should this name be based on the type of dispatcher?

File: libsplinter/src/network/dispatch.rs [hunk 6]
```diff
@@ -301,112 +295,149 @@ impl<MT: Hash + Eq + Debug + Clone> HandlerWrapper<MT> {
...
-            }
+    pub fn shutdown_signaler(&self) -> DispatchLoopShutdownSignaler<MT> {
+        DispatchLoopShutdownSignaler {
+            sender: self.sender.clone(),
         }
-        Ok(())
+    }
+}
+
+pub struct DispatchMessageSender<MT: Any + Hash + Eq + Debug + Clone> {
```
Comment (peterschwarz) [inline libsplinter/src/network/dispatch.rs:416]: Should this implement clone?

File: libsplinter/src/network/dispatch.rs (hunk 4 above)
Comment (agunde406) [reply libsplinter/src/network/dispatch.rs:326]: Done

File: libsplinter/src/network/dispatch.rs (hunk 5 above)
Comment (agunde406) [reply libsplinter/src/network/dispatch.rs]: I added an option to the builder to provide a thread name

File: libsplinter/src/network/dispatch.rs (hunk 6 above)
Comment (agunde406) [reply libsplinter/src/network/dispatch.rs:437]: Added

File: libsplinter/src/network/dispatch.rs [hunk 7]
```diff
@@ -301,112 +297,168 @@ impl<MT: Hash + Eq + Debug + Clone> HandlerWrapper<MT> {
...
+                    }
+                }
+            });
+
+        match join_handle {
+            Ok(join_handle) => Ok(DispatchLoop {
+                sender: tx,
+                join_handle,
+            }),
+            Err(err) => Err(format!("Unavle to start up dispatch loop thread: {}", err)),
```
Comment (peterschwarz) [inline libsplinter/src/network/dispatch.rs:401]: s/Unavle/Unable

Comment (agunde406) [reply libsplinter/src/network/dispatch.rs]: Fixed

**Pull request merged**